### PR TITLE
JDK-8298813: [C2] Converting double to float cause a loss of precision and resulting crypto.aes scores fluctuate

### PR DESCRIPTION
**Pull request by sunny868**
Hi all,
For C2, convert double to float cause a loss of precision,

<pre><code class="cpp">
./chaitin.cpp:221
_high_frequency_lrg = MIN2(double(OPTO_LRG_HIGH_FREQ), _cfg.get_outer_loop_frequency());
</code></pre>

Here, _high_frequency_lrg type is float, so maybe has a loss of precision. when it be used:

<pre><code class="cpp">
./coalesce.cpp:379
if( lrg._maxfreq >= _phc.high_frequency_lrg() ) {
   ...
}
</code></pre>
Here, lrg._maxfreq type is double, so _high_frequency_lrg will be convert double again. But now, due to the loss of precision of _high_frequency_lrg, the conditions here may be true or false.

There are two cases that I tested for SPECjvm2008 crypto.aes.
case 1:
<pre><code class="shell">
//chaitin.cpp:221
// fcvt.s.d $f0,$f0 #double->float
d = 16.994714324523816
f = 16.9947147

//coalesce.cpp:379
// fcvt.d.s $f0,$f0 #float->double
// fcmp.sle.d $fcc2,$f0,$f1
(gdb) i r fa0
fa0 {f = 0x0, d = 0x10} {f = -1.08420217e-19, d = 16.994714736938477}
(gdb) i r fa1
fa1 {f = 0x0, d = 0x10} {f = -7.68722312e-24, d = 16.994714324523816}
</code></pre>

case2:
<pre><code class="shell">
//chaitin.cpp:221
// fcvt.s.d $f0,$f0
d = 16.996332681816536
f = 16.9963322

//coalesce.cpp
// fcvt.d.s $f0,$f0
// fcmp.sle.d $fcc2,$f0,$f1
(gdb) i r fa0
fa0 {f = 0x0, d = 0x10} {f = -1.08420217e-19, d = 16.996332168579102}
(gdb) i r fa1
fa1 {f = 0x0, d = 0x10} {f = -1.73570044e-14, d = 16.996332681816536}
</code></pre>

The above two cases result in different block generation（case2 can insert new SpillCopyNodes), and resulting score on cryto.aes is fluctuate.

This is a patch to fix this problem. Please help review it.

Thanks,
Sun Guoyun

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8298813](https://bugs.openjdk.org/browse/JDK-8298813): [C2] Converting double to float  cause a loss of precision and resulting crypto.aes scores fluctuate

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**) ⚠️ Review applies to [f017e49c](https://git.openjdk.org/jdk/pull/11685/files/f017e49c74f6f2018e51aadae59b88829d469850)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11685/head:pull/11685` \
`$ git checkout pull/11685`

Update a local copy of the PR: \
`$ git checkout pull/11685` \
`$ git pull https://git.openjdk.org/jdk pull/11685/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11685`

View PR using the GUI difftool: \
`$ git pr show -t 11685`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11685.diff">https://git.openjdk.org/jdk/pull/11685.diff</a>

</details>

### CODE REVIEW

**Comment by sunny868** [comment]
The test FAIL and ERROR items has nothing to do with this patch.

File: src/hotspot/share/opto/chaitin.hpp [hunk 1]
```diff
@@ -478,7 +478,7 @@ class PhaseChaitin : public PhaseRegAlloc {
 
   Block **_blks;                // Array of blocks sorted by frequency for coalescing
 
-  float _high_frequency_lrg;    // Frequency at which LRG will be spilled for debug info
+  double _high_frequency_lrg;    // Frequency at which LRG will be spilled for debug info
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/chaitin.hpp:481]:
```suggestion
  double _high_frequency_lrg;   // Frequency at which LRG will be spilled for debug info
```

**Comment by sunny868** [comment]
OK. you can focus on the instruction generation of function com.sun.crypto.provider.DESCrypt::cipherBlock, sometimes it is 
<pre><code class="shell">
// result 63.4 ops/m
// B5->B112->B55

   B5: #   out( B112 B6 ) <- in( B4 )  Freq: 0.999996
    BRge   T0, A6, B112 #@branchConIU_reg_reg_short  P=0.000001 C=-1.000000
   ...
   B55: #  out( B56 ) <- in( B112 B53 B54 )  Freq: 3.54712e-05
    st_w    T6, [SP + #40]  # spill 9
    CALL,static #@CallStaticJavaDirect  wrapper for: uncommon_trap(reason='range_check' action='make_not_entrant' debug_id='0')
   ILLTRAP   ;#@ShouldNotReachHere
   ...
  B112: # out( B55 ) <- in( B5 )  Freq: 1.01327e-06
  mov    T6, #0 #@loadConI
  JMP    B55 #@jmpDir_short
</code></pre>
and sometimes it is
<pre><code class="shell">
// result: 41.99 ops/m
// B5->B55

     B5: #   out( B55 B6 ) <- in( B4 )  Freq: 0.999996
     mov    A6, #0 #@loadConI
     BRge   T8, T6, B55 #@branchConIU_reg_reg_short  P=0.000001 C=-1.000000

     B55: #  out( B56 ) <- in( B112 B53 B54 )  Freq: 3.54712e-05
             st_w    A6, [SP + #4]  # spill 9
     CALL,static #@CallStaticJavaDirect  wrapper for: uncommon_trap(reason='range_check' action='make_not_entrant' debug_id='0')
     ILLTRAP   ;#@ShouldNotReachHere
</code></pre>

**Comment by sunny868** [comment]
And in order to eliminate the fluctuation of running scores caused by C1, it is recommended that you use `-XX:-TieredCompilation`

**Comment by sunny868** [comment]
By the way, adding the following changes to this patch will allow crypto.aes to get a stable high score.
But I'm not sure if this change can be committed with this patch.
<pre><code class="shell">
diff --git a/src/hotspot/share/opto/coalesce.cpp b/src/hotspot/share/opto/coalesce.cpp
index b95987c4b09..82fbc72ad7e 100644                                           
--- a/src/hotspot/share/opto/coalesce.cpp                                       
+++ b/src/hotspot/share/opto/coalesce.cpp                                       
@@ -376,7 +376,7 @@ void PhaseAggressiveCoalesce::insert_copies( Matcher &matcher ) {
             LRG &lrg = lrgs(nidx);                                             
                                                                                
             // If this lrg has a high frequency use/def                        
-            if( lrg._maxfreq >= _phc.high_frequency_lrg() ) {                  
+            if( lrg._maxfreq > _phc.high_frequency_lrg() ) {                   
</code></pre>
But I'm not sure if this change can be committed with this patch or if a new patch can be committed for it.

**Comment by TobiHartmann** [comment]
Tests passed and performance results looks good. I think the other change that you proposed should not be part of this.

**Review by TobiHartmann — APPROVED** [review]

**Comment by TobiHartmann** [comment]
@sunny868 We take the average value of multiple test runs.

**Comment by sunny868** [comment]
> @sunny868 We take the average value of multiple test runs.
How many times do I need to run SPECjvm2008? And is this average from all jvm2008 benchmarks or a single one? 
Are there any documentation or websites about performance testing for my reference?

**Comment by TobiHartmann** [comment]
We run the SPECjvm2008 benchmarks individually and gather different results (average, low/high, statistical significance, ...). I think the exact configuration differs depending on the platform and benchmark. Since this is part of our internal performance testing infrastructure, no public documentation is available.

**Pull request closed without merging**